File: 098_Validate_Binary_Search_Tree/098_1.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.lower_upper(root, float("-inf"), float("inf"))

    def lower_upper(self, node: TreeNode, lower: int, upper: int) -> bool:
        if not node:
            return True
        val = node.val;
        if val <= lower or val >= upper:
            return False
        return self.lower_upper(node.left, lower, val) and \
               self.lower_upper(node.right, val, upper)


if __name__ == "__main__":
    node1 = TreeNode(5)
    node2 = TreeNode(1)
    node3 = TreeNode(4)
    node4 = TreeNode(3)
    node5 = TreeNode(6)
    node1.left = node2
    node1.right = node3
    node3.left = node4
    node3.right = node5
    s = Solution()
    print(s.isValidBST(node1))

File: 098_Validate_Binary_Search_Tree/098_1.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

long lastVal = LONG_MIN;

bool isValidBST(struct TreeNode * root) {
    if (!root) return true;
    if (!isValidBST(root->left))  return false;
    if (root->val <= lastVal) return false;
    lastVal = (long)root->val;
    if (!isValidBST(root->right)) return false;
    return true;
}

int main(void) {
    struct TreeNode *node1 = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    struct TreeNode *node2 = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    struct TreeNode *node3 = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    node1->val = 2; node2->val = 1; node3->val = 3;
    node1->left = node2; node1->right = node3;
    node2->left = NULL; node2->right = NULL; node3->left = NULL; node3->right = NULL;
    printf("%s\n", isValidBST(node1) ? "true" : "false");
    free(node1);
    free(node2);
    free(node3);
}

